hal/src/nRF52840/ble_hal.cpp

```diff
@@ -15,739 +15,3480 @@
-        d.conn_handle = event->evt.gatts_evt.conn_handle;
-        g_profile.callback(BLE_EVENT_DATA_SENT, &d, g_profile.userData);
-        break;
-    }
-    case BLE_GATTS_EVT_SYS_ATTR_MISSING: {
-        // No persistent system attributes
-        const uint32_t ret = sd_ble_gatts_sys_attr_set(event->evt.gatts_evt.conn_handle, nullptr, 0, 0);
-        if (ret != NRF_SUCCESS) {
-            LOG(ERROR, "sd_ble_gatts_sys_attr_set() failed: %u", (unsigned)ret);
-        }
-        break;
+    if (os_thread_create(&evtThread_, "BLE Event Thread", OS_THREAD_PRIORITY_NETWORK, bleEventDispatch, this, BLE_EVENT_THREAD_STACK_SIZE)) {
+        LOG(ERROR, "os_thread_create() failed");
+        os_queue_destroy(evtQueue_, nullptr);
+        return SYSTEM_ERROR_INTERNAL;
     }
-    case BLE_GATTS_EVT_TIMEOUT: {
-        // Disconnect on GATT server timeout
-        const uint32_t ret = sd_ble_gap_disconnect(event->evt.gatts_evt.conn_handle, BLE_HCI_REMOTE_USER_TERMINATED_CONNECTION);
-        if (ret != NRF_SUCCESS) {
-            LOG(ERROR, "sd_ble_gap_disconnect() failed: %u", (unsigned)ret);
+    return SYSTEM_ERROR_NONE;
+}
+
+int BleObject::BleEventDispatcher::enqueue(EventMessage& msg) {
+    if (os_queue_put(evtQueue_, &msg, 0, nullptr)) {
+        LOG(ERROR, "os_queue_put() failed.");
+        // Free the memory if needed.
+        if (msg.hook && msg.hookContext) {
+            msg.hook(&msg.evt, msg.hookContext);
         }
-        break;
+        return SYSTEM_ERROR_NO_MEMORY;
     }
-    case BLE_GATTC_EVT_TIMEOUT: {
-        // Disconnect on GATT client timeout
-        const uint32_t ret = sd_ble_gap_disconnect(event->evt.gattc_evt.conn_handle, BLE_HCI_REMOTE_USER_TERMINATED_CONNECTION);
-        if (ret != NRF_SUCCESS) {
-            LOG(ERROR, "sd_ble_gap_disconnect() failed: %u", (unsigned)ret);
+    return SYSTEM_ERROR_NONE;
+}
+
+void BleObject::BleEventDispatcher::onGenericEventCallback(on_ble_evt_cb_t cb, void* context) {
+    BleGenericEventHandler evtHandler;
+    evtHandler.handler = cb;
+    evtHandler.context = context;
+    genericEventHandlers_.append(evtHandler);
+}
+
+os_thread_return_t BleObject::BleEventDispatcher::bleEventDispatch(void* param) {
+    BleEventDispatcher* dispatcher = static_cast<BleEventDispatcher*>(param);
+    while (1) {
+        EventMessage msg;
+        if (!os_queue_take(dispatcher->evtQueue_, &msg, CONCURRENT_WAIT_FOREVER, nullptr)) {
+            if (msg.handler) {
+                // Invoke the handler registered by specific application.
+                if (msg.evt.type == BLE_EVT_SCAN_RESULT) {
+                    msg.handler(&msg.evt.params.scan_result, msg.context);
+                } else if (msg.evt.type == BLE_EVT_SVC_DISCOVERED) {
+                    msg.handler(&msg.evt.params.svc_disc, msg.context);
+                } else if (msg.evt.type == BLE_EVT_CHAR_DISCOVERED) {
+                    msg.handler(&msg.evt.params.char_disc, msg.context);
+                }
```

Comment:
That in some degree makes sense, while since the SoftDevice events are dispatched to several objects, like `Broadcaster`, 'Observer' etc, there will have lots of meaningless log printed. Logging the unhandled events in just one object would make the log more clear.